darts/dataprocessing/transformers/window_transformer.py

```diff
@@ -0,0 +1,691 @@
+                        getattr(pd.DataFrame(), BUILTIN_TRANSFORMS[fn][0].__name__)(
+                            **function_group_available_kwargs
+                        ),
+                        fn,
+                    ).__code__.co_varnames
+                )
+                function_available_keys = list(
+                    function_expected_args.intersection(set(keys))
+                )
+                function_available_kwargs = {
+                    k: v
+                    for k, v in transformation.items()
+                    if k in function_available_keys
+                }
+
+                return function_group_available_kwargs, function_available_kwargs
+            elif callable(fn):
+                # if user function, get the kwargs for the function
+                function_available_kwargs = {
+                    k: v for k, v in transformation.items() if k in keys
+                }
+                return function_available_kwargs
+            else:
+                raise_log(
+                    Exception("The transformation function is not valid."), logger
+                )
+
+        # get the kwargs from the transform() function call.
+        fillnamethod = kwargs.get(
+            "fillnamethod",
+            "bfill"
+            if (
+                ("navalue" not in kwargs)
+                or ("navalue" in kwargs and kwargs["navalue"] is None)
+            )
+            else None,
+        )
+        navalue = kwargs.get("navalue", None)
+
+        BUILTIN_TRANSFORMS = builtins
+
+        fn = transformation["function"]
+
+        df_ts = series.pd_dataframe(
+            copy=True
+        )  # get the series values in a dataframe. TODO: check if copy is necessary
+
+        if isinstance(fn, str):
+            # verification of the value of the string should have been already validated in the constructor
+
+            function_group = BUILTIN_TRANSFORMS[fn][0].__name__
+            function_name = BUILTIN_TRANSFORMS[fn][1]
+
+            # if rolling, we need to run through window list
+            if function_group == "rolling":
+                transf_ts = []  # list to get all windows resulting transformations
+                for window in transformation["window"]:
+                    copy_transformation = copy.deepcopy(
+                        transformation
+                    )  # to avoid writing the original dict
+                    copy_transformation["window"] = window
+                    # get function_group and function kwargs
+                    function_group_kwargs, function_kwargs = _get_function_kwargs(
+                        copy_transformation, BUILTIN_TRANSFORMS
+                    )
+
+                    if "closed" not in function_group_kwargs:
+                        function_group_kwargs[
+                            "closed"
+                        ] = "left"  # to garantee that the latest value is not included in the window: forecasting safe
```

Comment:
How about when "closed" is in the kwargs? Would this still be "forecasting safe"?

Reply:
if the user provides a value for closed that is not "left" then no, we cannot guarantee forecasting safe. Should we warn the user about this if they provide the "closed" item? We can give the user a warning, but the computation can still go through, or we can raise an Error and block the computation. I guess we have to always provide forecasting safe behaviour, correct?

Reply:
I added user warning and forecasting_safe enforcement in TimeSeries.window_transform(). The default configuration of the function is to be forecasting safe